sycl/doc/SYCL2020-SpecializationConstants.md

```diff
@@ -0,0 +1,766 @@
+  scalar value within the composite, i.e. the position where scalar
+  specialization constant resides within the byte array supplied by the user.
+  For scalar specialization constants it will always be 0.
+- Size of the scalar specialization constant
+
+For example, the following code:
+```
+struct Nested {
+  float a, b;
+};
+struct A {
+  int x;
+  Nested n;
+};
+
+specialization_id<int> id_int;
+specialization_id<A> id_A;
+// ...
+  [=](kernel_handler h) {
+    h.get_specialization_constant<id_int>();
+    h.get_specialization_constant<id_A>();
+  }
+```
+
+Will result in the following property set generated:
+```
+property_set {
+  Name = "SYCL/specialization constants",
+  properties: [
+    property {
+      Name: "id_int_symbolic_ID",
+      ValAddr: points to byte array [{0, 0, 4}],
+      Type: PI_PROPERTY_TYPE_BYTE_ARRAY,
+      Size: sizeof(byte array above)
+    },
+    property {
+      Name: "id_A_symbolic_ID",
+      ValAddr: points to byte array [{1, 0, 4}, {2, 4, 4}, {3, 8, 4}],
+      Type: PI_PROPERTY_TYPE_BYTE_ARRAY,
+      Size: sizeof(byte array above)
+    },
+  ]
+}
+```
+
+#### DPC++ runtime
+
+For each device binary compiler generates a map
+\<Symbolic ID\> =\> \<list of spec constant descriptors\> ("ID map"). DPC++
+runtime imports that map when loading device binaries.
+It also maintains another map \<Spec const symbolic ID\> =\> \<its value\> 
+("value map") per `sycl::kernel_bundle` object. The value map is updated upon
+`kernel_bundler::set_specialization_constant<ID>(val)` and
+`handler::set_specialization_constant<ID>(val)` calls from the app.
+
+In order for runtime to access the right property, it need to compute the
+symbolic ID of a specialization constant based on user-provided inputs, such
+as non-type template argument passed to `set_specialization_constant` argument.
+DPC++ Headers section describes how symbolic IDs are generated and the same
+trick is used within `set_specialization_constant` method:
+```
+template<auto& SpecName>
+void set_specialization_constant(
+  typename std::remove_reference_t<decltype(SpecName)>::type value) {
+  const char *SymbolicID =
+#if __has_builint(__builtin_unique_stable_name)
+      __builtin_unique_stable_name(detail::specialization_id_name_generator<S>);
+#else
+      // without the builtin we can't get the symbolic ID of the constant
+      "";
+#endif
+  // remember the value of the specialization constant
+  SpecConstantValuesMap[SymbolicID] = value;
+}
+```
+
+The major downside of that approach is that it can't be used with any
+third-party host compiler, because it uses a specific built-in function to
+generate symbolic IDs of specialization constants. Good solution would be to
+employ integration header here, i.e. we could provide some class template
+specializations which will return symbolic IDs - the same approach as we use
+for communicating OpenCL kernel names from the compiler to the runtime.
```

Comment:
It is a requirement that dpc++ input should be consumable by a third party host compiler.  So, please design such that this requirement is satisfiable.

Reply:
This [discussion](https://github.com/intel/llvm/pull/3331/files#r593461133) below addresses this concern. The plan is that the first implementation will rely on the above DPC++-specific features, later one of the host compiler-neutral scheme suggested in the discussion will be implemented.

Reply:
That will mean that any source code that contains a kernel function that uses a specialization constant will not be compilable with a third party compiler in the first implementation.

So, in the release notes, we would have to put that specialization constants are only supported in modules compiled by the dpcpp compiler.

I am not sure that meets requirements, and may need a special waiver from oneAPI PXT.

Reply:
Is it possible to continue the discussion in https://github.com/intel/llvm/pull/3331#discussion_r595561411 ?